### PR TITLE
add last_active info to cluster list

### PR DESCRIPTION
**Pull request by BelSasha**

### CODE REVIEW

**Comment by BelSasha** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1231?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1231** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1231?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1227** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1225** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/1225?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @BelSasha and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: runhouse/main.py [hunk 1]
```diff
@@ -669,27 +660,45 @@ def cluster_list():
...
+
+        # currently relying on status pings to den as a sign of cluster activity.
+        # The split is required to remove milliseconds and the offset (according to UTC) from the timestamp.
+        # (status_last_checked is in the following format: YYYY-MM-DD HH:MM:SS.ssssss±HH:MM)
+
+        last_active_at = den_cluster.get("status_last_checked")
+        last_active_at = (
+            datetime.datetime.fromisoformat(last_active_at.split(".")[0])
+            if isinstance(last_active_at, str)
+            else datetime.datetime(1970, 1, 1)
```
Comment (jlewitt1) [inline runhouse/main.py:672]: why use this date if not found?

File: runhouse/main.py [hunk 2]
```diff
@@ -669,27 +660,45 @@ def cluster_list():
...
+
+        if cluster_status == "running" and not last_active_at:
+            # For BC, in case there are clusters that were saved and created before we introduced sending cluster status to den.
+            cluster_status = "unknown"
+
+        cluster_info = {
+            "Name": cluster_name,
+            "Cluster Type": cluster_type,
+            "Status": cluster_status,
+            "Last Active (UTC)": last_active_at,
```
Comment (jlewitt1) [inline runhouse/main.py:684]: do you have a screenshot you could share for this PR similar to the previous one showing how it would look in the CLI?

Comment (BelSasha) [reply runhouse/main.py:684]: image

File: runhouse/main.py [hunk 3]
```diff
@@ -702,16 +716,25 @@ def cluster_list():
...
-        table.add_row(
-            rh_cluster.get("Name"),
-            rh_cluster.get("Cluster Type"),
-            StatusColors.get_status_color(rh_cluster.get("Status")),
-        )
+        last_active_at = rh_cluster.get("Last Active (UTC)")
+
+        # Print Running clusters that were active it the last 24 hours
+        now = datetime.datetime.utcnow().replace(tzinfo=datetime.timezone.utc)
+        if (now - last_active_at).total_seconds() <= LAST_ACTIVE_AT_TIMEFRAME:
```
Comment (jlewitt1) [inline runhouse/main.py:729]: wasnt so clear in the CLI output that only running clusters active in last 24 hrs are being shown, can we make that more clear in the table somehow?

**Review by jlewitt1 — APPROVED** [review]

**Pull request merged**